feat(CI/CD): Add CI/CD documentation (#1922)
mendersoftware/mender-docs: 06.Artifact-creation/10.CI-CD/docs.md

---
title: CI/CD pipelines
taxonomy:
    category: docs
    label: tutorial
---

## CI/CD pipelines

Mender provides support for application updates in the form of CI/CD pipelines to implement all kinds of specific use-cases. This section describes possible ways to integrate your workflow with Mender and most popular CI/CD providers.

### Mender CI-workflows docker image

> To help you in creating CI/CD pipelines, we created Github repository containing ready to use, battle-tested CI/CD pipeline templates that can be used your workflows. Additionally, we provide docker image with command-line utilities,`mender-artifact` and `mender-cli` installed.

Reviewer: Docker

Author: Done!

> To help you in creating CI/CD pipelines, we created Github repository containing ready to use, battle-tested CI/CD pipeline templates that can be used your workflows. Additionally, we provide docker image with command-line utilities,`mender-artifact` and `mender-cli` installed.

Reviewer: What is the name of the Docker image? We should have it (with a link to Docker Hub)

Author: Done, thanks!

> To help you in creating CI/CD pipelines, we created Github repository containing ready to use, battle-tested CI/CD pipeline templates that can be used your workflows. Additionally, we provide docker image with command-line utilities,`mender-artifact` and `mender-cli` installed.

Reviewer: We should also add a note in the Downloads sections of `mender-artifact` and `mender-cli` to inform about this new Docker image

Author: Done, thanks!

> Location: https://github.com/mendersoftware/mender-ci-workflows

Reviewer: Looks a bit cryptic. What about something like "The CI/CD templates can be found at mender-ci-workflows repository" (with a link to the GitHub repo)

### Supported CI/CD providers

Mender supports all major CI/CD providers:

* GitHub Actions

> * Gitlab

Reviewer: `GitLab` or `GitLab CI/CD`

Author: Done!

> * GitHub Actions
> * Gitlab
> * Azure DevOps
> * buildbot
> * Bamboo
> * Travis
> * Circle CI

Reviewer: Let's write for now only the three major ones that we are committed to support
```suggestion
* GitHub Actions
* Gitlab
* Azure DevOps
```

and we can add more as we go

Reviewer: Let's also create the subpages for these three (empty content for now) so that we can link them from here.

Author: Done, thanks!

> ### Signing an Mender Artefact

Reviewer: Here and below
```suggestion
### Signing an Mender Artifact
```

Author: Done!

To sign your artifact, you can use `mender-artifact` command-line utility embedded into docker image

> More details about signing Mender Artifacts: https://docs.mender.io/artifact-creation/sign-and-verify

Reviewer: These links need to be proper markdown syntax.

Author: Done!

### Uploading an Mender Artefact 

> 

Reviewer: The direction is good. The use cases need to be well described not assuming much knowledge of Mender ecosystem.

Author: Thanks! Please let me know if the current docs are satisfying for you, then I'll create feature branch.

Reviewer: No, sorry. We need a bit more of an introduction to each scenario.. See comment for `Creating and Signing an Mender Artifact` and let's do something similar to the rest.

---
title: CI/CD pipelines
taxonomy:
    category: docs

>     label: tutorial

Reviewer: This is not really a tutorial. Remove this line.

Author: Done!

---
title: CI/CD pipelines
taxonomy:
    category: docs
    label: tutorial
---

## CI/CD pipelines

Mender provides support for application updates in the form of CI/CD pipelines to implement all kinds of specific use-cases. This section describes possible ways to integrate your workflow with Mender and most popular CI/CD providers.

### Mender CI-workflows Docker image

> To help you in creating CI/CD pipelines, we created Github repository containing ready to use, battle-tested CI/CD pipeline templates that can be used your workflows. Additionally, we provide [Docker image](https://hub.docker.com/r/mendersoftware/mender-ci-tools) with command-line utilities,`mender-artifact` and `mender-cli` installed.

Reviewer: Missing a link to the repository.

"... we created **a** [Github repository](https://github.com/mendersoftware/mender-ci-workflows/) containing..."

Also, let's be more generic with:

"battle-tested CI/CD pipeline code snippets that can be used your workflows"

(let's avoid "actions" or "templates" keywords here)

Reviewer: "We provide mender-ci-tools Docker image based on Alpine Linux and providing command-line utilities...."

Author: Done!

> In order to pull the image, you need to issue following command:

Reviewer: ```suggestion
To locally inspect the Docker image, pull it with the following command:
```

Author: Done!

<!--AUTOVERSION: "mendersoftware/mender-ci-tools:%"/ignore-->
```
<!--AUTOVERSION: "mendersoftware/mender-ci-tools:%"/ignore-->
docker pull mendersoftware/mender-ci-tools:master
```

The Docker image definition and CI/CD templates can be found [here](https://github.com/mendersoftware/mender-ci-workflows).

### Supported CI/CD providers

Mender currently supports three major CI/CD providers:

* [GitHub Actions](./01.GitHub%20Actions/docs.md)
* [GitLab CI/CD](./02.GitLab/docs.md)
* [Azure DevOps](./03.Azure%20DevOps/docs.md)

### Creating and Signing an Mender Artifact

> To create and sign your artifact, you can use `mender-artifact` command-line utility embedded into Docker image.
> 
> You can read about artifact creation details [here](../01.Create-an-Artifact/docs.md#create-a-full-filesystem-update-artifact).
> 
> More details about signing Mender Artifacts [here](https://docs.mender.io/artifact-creation/sign-and-verify).

Reviewer: ```suggestion
This scenario uses `mender-artifact` tool to package your software as a Mender Artifact. A Mender Artifact can optionally be signed, so the examples provided allow for signing the Artifact too.

The input for the scenario is the software being packaged and the output is a Mender Artifact.

You can read about artifact creation details [here](../01.Create-an-Artifact/docs.md#create-a-full-filesystem-update-artifact).

More details about signing Mender Artifacts [here](https://docs.mender.io/artifact-creation/sign-and-verify).
```

Reviewer: Please add something similar in the other scenarios, mentioning what is the expected input and output.

Author: Done plase let me know how `Deploying an Artifact to the fleet of devices` section should look like: given various CI/CD providers, there's no single solution here.

Author: Pushed changes after chat discussion, thanks!
